src/flb_utils.c

```diff
@@ -732,20 +743,102 @@ int flb_utils_write_str(char *buf, int *off, size_t size,
             i += (hex_bytes - 1);
         }
         else if (c > 0xFFFF) {
-            hex_bytes = flb_utf8_len(str + i);
-            if (available - written < 6) {
-                return FLB_FALSE;
-            }
+            utf_sequence_length = flb_utf8_len(str + i);
 
-            if (i + hex_bytes > str_len) {
+            if (i + utf_sequence_length > str_len) {
                 break; /* skip truncated UTF-8 */
             }
-            for (b = 0; b < hex_bytes; b++) {
-                tmp[b] = str[i+b];
+
+            is_valid = FLB_TRUE;
+            for (utf_sequence_number = 0; utf_sequence_number < utf_sequence_length;
+                utf_sequence_number++) {
+                /* Leading characters must start with bits 11 */
+                if (utf_sequence_number == 0 && ((str[i] & 0xC0) != 0xC0)) {
```

Comment:
This validation is too late, in my opinion. It should be done at the time the sequence length is calculated. How is a valid length being returned if the leading byte is invalid?

Reply:
If the leading byte is invalid, we need to process this leading byte as 1 invalid byte which is mapped to an our private use area character.

Since we do not throw away bytes, we still need to list out a sequence. When the sequence is evaluated to be invalid, we switch to denoting it as a list of mapped private use area characters rather than a valid unicode character. There are test cases for both options, invalid leading and trailing bytes.